src/libraries/System.Private.Xml/src/System/Xml/Xsl/XPath/XPathBuilder.cs

```diff
@@ -109,17 +113,21 @@ public virtual QilNode Number(double value)
             return _f.Double(value);
         }
 
-        public virtual QilNode Operator(XPathOperator op, QilNode left, QilNode right)
+        public virtual QilNode Operator(XPathOperator op, [AllowNull] QilNode left, [AllowNull] QilNode right)
         {
             Debug.Assert(op != XPathOperator.Unknown);
-            switch (s_operatorGroup[(int)op])
+            XPathOperatorGroup opGroup = s_operatorGroup[(int)op];
+
+            Debug.Assert((opGroup != XPathOperatorGroup.Negate && right != null) || (opGroup == XPathOperatorGroup.Negate && right == null));
```

Comment:
Why `XPathOperatorGroup.Union` should always receive `left` as `null`? 

Also, if I change the assertion to what you are suggesting, any `XPathOperatorGroup` other than `Union` and `Negate` would fail.